repos/system_upgrade/el7toel8/actors/migratentp/libraries/ntp2chrony.py

```diff
@@ -0,0 +1,669 @@
+            else:
+                return False
+
+        self.trusted_keys = key_ranges
+        return True
+
+    def write_chrony_configuration(self, chrony_conf_path, chrony_keys_path,
+                                   dry_run=False, backup=False):
+        chrony_conf = self.get_chrony_conf(chrony_keys_path)
+        logging.debug("Generated %s:\n%s", chrony_conf_path, chrony_conf)
+
+        if not dry_run:
+            self.write_file(chrony_conf_path, 0o644, chrony_conf, backup)
+
+        chrony_keys = self.get_chrony_keys()
+        if chrony_keys:
+            logging.debug("Generated %s:\n%s", chrony_keys_path, chrony_keys)
+
+        if not dry_run:
+            self.write_file(chrony_keys_path, 0o640, chrony_keys, backup)
+
+    def get_processed_time_sources(self):
+        # Convert {0,1,2,3}.*pool.ntp.org servers to 2.*pool.ntp.org pools
+
+        # Make shallow copies of all sources (only type will be modified)
+        time_sources = [s.copy() for s in self.time_sources]
+
+        pools = {}
+        for source in time_sources:
+            if source["type"] != "server":
+                continue
+            m = re.match("^([0123])(\\.\\w+)?\\.pool\\.ntp\\.org$", source["address"])
+            if m is None:
+                continue
+            number = m.group(1)
+            zone = m.group(2)
+            if zone not in pools:
+                pools[zone] = []
+            pools[zone].append((int(number), source))
+
+        remove_servers = set()
+        for zone, pool in pools.items():
+            # sort and skip all pools not in [0, 3] range
+            pool.sort()
+            if [number for number, source in pool] != [0, 1, 2, 3]:
+                # only exact group of 4 servers can be converted, nothing to do here
+                continue
+            # verify that parameters are the same for all servers in the pool
+            if not all([p[1]["options"] == pool[0][1]["options"] for p in pool]):
+                break
+            remove_servers.update([pool[i][1]["address"] for i in [0, 1, 3]])
+            pool[2][1]["type"] = "pool"
+
+        processed_sources = []
+        for source in time_sources:
+            if source["type"] == "server" and source["address"] in remove_servers:
+                continue
+            processed_sources.append(source)
+        return processed_sources
+
+    def get_chrony_conf_sources(self):
+        conf = ""
+
+        if self.step_tickers:
+            conf += "# Specify NTP servers used for initial correction.\n"
+            conf += "initstepslew 0.1 {}\n".format(" ".join(self.step_tickers))
+            conf += "\n"
+
+        conf += "# Specify time sources.\n"
+
+        for source in self.get_processed_time_sources():
+            address = source["address"]
+            if address.startswith("127.127."):
+                if address.startswith("127.127.1."):
+                    continue
+                assert False
```

Comment:
Sorry, not getting it - what is this assert used for? What special case are you addressing by it?

Reply:
+1, what would make more sense to me:
```
if not address.startswith('127.127.1.'):
   raise Exception('Some meaningful message as to why this is not correct')
```

Reply:
asserts catch bugs in the code. In a bug-free code they do nothing. In this case there are two places in the code that need to be consistent with each other. The parser and the config generator. If the parser was modified to handle another refclock, but the config generator wasn't, the assert would trigger.

Reply:
@mlichvar This is handling user provided data, there may not be a bug in the code, but the data can be ill-formed, and you NEVER use asserts for that. Please also read the guidance about "In-your-face error handling", the point here is that these things MAY bubble up to users, and hence the very clear distinction between:
```
AssertionError: False
```
And
```
NtpMigrationError: Configuration in /some/path contains multiple refclocks
```

Reply:
But that doesn't apply to this case. No user provided data can trigger that assert. The parser ignores all refclocks except 127.127.1.0, so they are listed in the generated file as ignored directives. The assert will trigger only if the parser is updated to handle other refclocks, but the generator is not, i.e. there is a bug in the logic of the code.

Reply:
Thanks for the clarification @mlichvar - in this case I'd suggest removing the two checks altogether and flesh them out into a unit test, because that's what it is, we will not be testing if someone made an incompatible change to he parser when running on a production system, we need to test for that before we release.

As a side note, about asserts, since I'm the one who wrote that section of the guidelines, the "mutating assert" scenario is spelled out merely as a worst case scenario, but the more fundamental takeaway is that relying on assert is simply making assumptions as to how people are executing our code. We cannot make these assumptions.

Reply:
Right, there should be tests with non-LOCAL refclocks. And those may or may not trigger the assert depending on what bug the code actually has. Checking the output is important, but it may not catch everything.